QPOptions (#5169)
knative/docs: docs/serving/services/using-queue-extensions.md

# Using extensions enabled by QPOptions

> Once your cluster is setup with extensions enabled by QPOptions, a Service can decide which extensions it wish to use and how to configure such extensions. Activating and configuring extensions is described here.

Reviewer: It still doesn't explain what QPOptions is here - I would like to see a proper conceptual intro so that users know why they would enable this or what they would use it for. For example, something like:

"QPOptions is an optional Go package that provides extensions for Knative Services. These extensions allow you to..." etc

I don't think what's currently here is enough to warrant adding a new section to the docs.

Author: Added an intro to what is QPOptions

> Once your cluster is setup to allow services to use Queue Proxy Extensions, a service can decide which extensions it wish to use and how to configure such extensions. Activating and configuring extensions is described here.

Reviewer: Please refer to the docs style guide re formatting for things like Services https://github.com/knative/docs/blob/main/contribute-to-docs/style-guide/style-and-formatting.md

> ## Prerequisites for preparing the cluster

Reviewer 2: Probably a good idea to move this section down so that the prerequisites are grouped together.

The Prereqs could be in one section, for example: 

```
## Prerequisites

Before you can use extensions enabled by QPoptions, you must:

- Prepare your cluster:
  - Make sure you are using a Queue Proxy image...
  - Make sure that the cluster config-features is set with...
- Meet the prerequisites in [Creating a Service](...)
```

> 1. Make sure you are using a Queue Proxy image that was built with the extensions that you wish to use - See [Extending Queue Proxy image with QPoptions](../queue-extensions.md).

Reviewer 2: ```suggestion
1. Make sure you are using a Queue Proxy image that was built with the extensions that you wish to use. See [Extending Queue Proxy image with QPoptions](../queue-extensions.md).
```

> 1. Make sure that the cluster config-features is set with: `queueproxy.mount-podinfo: allowed`  - See [Enabling Queue Proxy Pod Info](../configuration/feature-flags.md#queue-proxy-pod-info) for more details.

Reviewer 2: ```suggestion
1. Make sure that the cluster config-features is set with `queueproxy.mount-podinfo: allowed`. See [Enabling Queue Proxy Pod Info](../configuration/feature-flags.md#queue-proxy-pod-info) for more details.
```

## Overview

> A service can activate and configure queue proxy extensions by adding `qpextension.knative.dev/*` annotations under the: `spec.template.metadata` of the service custom resource definition.

Reviewer: same comment re formatting for Service

> A service can activate and configure queue proxy extensions by adding `qpextension.knative.dev/*` annotations under the: `spec.template.metadata` of the service custom resource definition.

Reviewer: re queue proxy extensions use capitalization and naming consistently throughout docs please

> Setting a value of: `qpoption.knative.dev/<ExtensionName>-activate: "enable"` will activate the extension.

Reviewer 2: ```suggestion
Setting a value of: `qpoption.knative.dev/<ExtensionName>-activate: "enable"` activates the extension.
```

> Setting a value of: `qpoption.knative.dev/<ExtensionName>-config-<Key>: "<Val>"` will add a configuration of Key: Val to the extension.

Reviewer 2: ```suggestion
Setting a value of: `qpoption.knative.dev/<extension-name>-config-<key>: "<value>"` adds a configuration of `key: value` to the extension.
```

> In addition, the Service need to ensure that the Pod Info volume is mounted by adding the `features.knative.dev/queueproxy-podinfo: enabled` annotation under the: `spec.template.metadata` of the Service CRD.

Reviewer 2: ```suggestion
In addition, the Service must ensure that the Pod Info volume is mounted by adding the `features.knative.dev/queueproxy-podinfo: enabled` annotation under the: `spec.template.metadata` of the Service CRD.
```

          template:
            metadata:
                annotations:
                  features.knative.dev/queueproxy-podinfo: enabled
                  qpoption.knative.dev/testgate-activate: enable
                  qpoption.knative.dev/testgate-config-response: CU
                  qpoption.knative.dev/testgate-config-sender: Joe
            spec:
              containers:
                - image: gcr.io/knative-samples/helloworld-go
                  env:
                    - name: TARGET
                      value: "World"
        ```

    1. Apply the YAML file by running the command:

        ```bash
        kubectl apply -f <filename>.yaml
        ```
        Where `<filename>` is the name of the file you created in the previous step.

=== "kn CLI"

    ```
    kn service create helloworld-go \
        --image gcr.io/knative-samples/helloworld-go \
        --env TARGET=World \
        --annotation features.knative.dev/queueproxy-podinfo=enabled \
        --annotation qpoption.knative.dev/testgate-activate=enable \
        --annotation qpoption.knative.dev/testgate-config-response=Goodbye \
        --annotation qpoption.knative.dev/testgate-config-sender=Joe
    ```

> After the Service has been created, Knative propagates the annotations to the podSpec of the Service deployment. When a Service pod is created, the Queue Proxy sidecar will mount a volume that contains the pod annotations and activate the `testgate` extension (if such an extension is available in the Queue Proxy image). The `testgate` extension will than be configured with the configuration: `{ sender: "Joe", response: "CU"}`.

Reviewer 2: ```suggestion
After the Service has been created, Knative propagates the annotations to the podSpec of the Service deployment. When a Service pod is created, the Queue Proxy sidecar will mount a volume that contains the pod annotations and activate the `testgate` extension. This occurs if the `testgate` extension is available in the Queue Proxy image. The `testgate` extension will then be configured with the configuration: `{ sender: "Joe", response: "CU"}`.
```
